### PR TITLE
Supply OnKeyPress with modifier-affected chars in mac

### PR DESCRIPTION
**Pull request by plankp**
Fixes https://github.com/ppy/osu/issues/3120

### CODE REVIEW

**Comment by swoolcock** [comment]
Have you tested this?  I can just about guarantee it will break for many modifier combinations, as it applies them to the text before returning the actual pressed key.

Have you read the documentation of `charactersIgnoringModifiers`?

https://developer.apple.com/documentation/appkit/nsevent/1524605-charactersignoringmodifiers

> This property is set to the non-modifier key character pressed for dead keys, such as Option-e. For example, Option-e (no shift key) returns an “e" for this method, whereas the characters property returns an empty string.

If you press option-e you're going to get a `flagsChanged` event for the option key, and have no corresponding "e" `keyDown`.

**Comment by plankp** [comment]
This only handles the caps lock modifier. Also, I tested Option-E on my mac. It still gives an empty string prior to this change.

For the option modifier, I cannot quite figure out a way to implement it.

**Comment by plankp** [comment]
That is why I opened the option keys as a separate issue, https://github.com/ppy/osu/issues/3125

**Comment by swoolcock** [comment]
I just made a simple test app in Xcode, overrode these functions in `NSWindow`, and got the following:

```
- (void)keyDown:(NSEvent *)event {
    NSLog(@"keydown: characters=%@, charactersIgnoringModifiers=%@", event.characters, event.charactersIgnoringModifiers);
}

- (void)keyUp:(NSEvent *)event {
    NSLog(@"keydown: characters=%@, charactersIgnoringModifiers=%@", event.characters, event.charactersIgnoringModifiers);
}

- (void)flagsChanged:(NSEvent *)event {
    NSLog(@"flagschanged: flags=%lu", event.modifierFlags);
}
```

Press option:

`2018-07-27 09:30:30.854019+0930 TestApp2[16790:32773575] flagschanged: flags=524576`

Release option:

`2018-07-27 09:30:30.933752+0930 TestApp2[16790:32773575] flagschanged: flags=256`

Press e:

`2018-07-27 09:30:32.890056+0930 TestApp2[16790:32773575] keydown: characters=e, charactersIgnoringModifiers=e`

Release e:

`2018-07-27 09:30:33.021633+0930 TestApp2[16790:32773575] keyup: characters=e, charactersIgnoringModifiers=e`

Press option:

`2018-07-27 09:30:34.245883+0930 TestApp2[16790:32773575] flagschanged: flags=524576`

Press e:

`2018-07-27 09:30:35.197804+0930 TestApp2[16790:32773575] keydown: characters=, charactersIgnoringModifiers=e`

Release e:

`2018-07-27 09:30:35.333735+0930 TestApp2[16790:32773575] keyup: characters=, charactersIgnoringModifiers=e`

Release option:

`2018-07-27 09:30:35.909825+0930 TestApp2[16790:32773575] flagschanged: flags=256`

**Comment by plankp** [comment]
Does typing cause the correct letter to display on text box?

**Comment by plankp** [comment]
Also, if that is a concern, process events could just query both characters and with modifiers ignored. Pass string from characters if it is non-empty and use with modifiers ignored as fallback.

**Comment by swoolcock** [comment]
This is just a blank `NSWindow` capturing key presses.  I haven't tried in osu! because I'm at work.

When generating key events, we only care about hardware keypresses.  This is why it uses `charactersIgnoringModifiers`, because it will not modify the character based on other keys.

**Comment by plankp** [comment]
I understand that.

However, in the code below, it passes the individual characters to `OnKeyPress` which is then handled by some pathway that I do not remember, through the ITextInput thing and textbox consumes that char combined text directly and pipes it onto the screen.

Based on that pathway, I think it will be best if we had OnKeyPress handle characters that were already affected by the modifier keys (only caps lock and shift in this case, ignore modifier still takes shift into account)

`KeyPressed` (or some name along the lines) is a method overrided by various components. The keychar it provides is supposed to have taken the caps lock stuff into account, which is why I changed it this way.

**Comment by swoolcock** [comment]
If you can make the change in a way that doesn't break the input manager visual tests in osu!framework, go for it. 👍

**Comment by plankp** [comment]
By manually copying the dll file into the osu.Framework.Test's output directory, this allows caps lock to work in text boxes. It does not resolve the option-based keys; they were already broken (as in, if you pressed option-e, `e` does not appear on screen as `charactersIgnoreModifiers` promised)

**Comment by swoolcock** [comment]
`charactersIgnoringModifiers` is working as expected. The issue is that in macOS we are dropping all key presses in the text input source if modifier keys are being held.

https://github.com/ppy/osu-framework/blob/master/osu.Framework/Input/GameWindowTextInput.cs

**Comment by plankp** [comment]
Is this for caps lock? or the option key?

**Comment by swoolcock** [comment]
The option key.  Pressing option+e will capture both the option and e presses in the input manager, but the text input source will ignore it.

**Comment by plankp** [comment]
Ok, so for both to work, I would use `charactersIgnoringModifiers` as a fallback, and do a separate fix in the osu-framework to resolve the option key?

**Comment by swoolcock** [comment]
A preferred solution:

1) Move the `flagsModified` hack from osu!framework into `CocoaNativeWindow` so that it sends modifiers as keypresses.
2) Change osu!framework to send typed keystrokes (with modifiers applied) to the text input source, rather than raw key events.  Not sure how you're going to do this at a platform-independent level.

A hacky solution:

1) Add caps lock to the `flagsModifed` handler in osu!framework.
2) Change osu!framework to handle typed keystrokes, as above.

**Comment by plankp** [comment]
for 2nd preferred solution, how would it deal with accented option keys? It takes two key presses to apply them. (Also, is this not how windows processes the caps locks modifier?). I don't think I understand the `flagsModified` hack thing in 1st point. Is this about the closed PR in framework? (I believe ekrctb wants me to ignore that one)

**Comment by swoolcock** [comment]
Fixing caps lock and accented key input are related issues and can likely be done in one PR in osu!framework, or two small ones in OpenTK and osu!framework.

**Comment by plankp** [comment]
There, revert back to using `charactersIgnoreModifiers`. Checking for caps lock status on `KeyDown` instead because `FlagsChanged` lost some caps lock updates when I switched windows.

Anything else I should change?

File: src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs [hunk 1]
```diff
@@ -653,15 +655,15 @@ public override void ProcessEvents()
...
+                            var s = Cocoa.FromNSString(Cocoa.SendIntPtr(e, selCharactersIgnoringModifiers));
                             foreach (var c in s)
                             {
                                 int intVal = (int)c;
                                 if (!Char.IsControl(c) && (intVal < 63232 || intVal > 63235))
                                 {
                                     // For some reason, arrow keys (mapped 63232-63235)
                                     // are seen as non-control characters, so get rid of those.
-                                    OnKeyPress(c);
+                                    OnKeyPress(capslockOn ? char.ToUpper(c) : c);
```
Comment (swoolcock) [inline src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs]: 1) `OnKeyPress` is intended to map to hardware keys, not characters, so you shouldn't be altering the value passed into `OnKeyPress`.
2) `ToUpper` does not take into account localisation.

File: src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs [hunk 2]
```diff
@@ -678,6 +680,8 @@ public override void ProcessEvents()
                     case NSEventType.FlagsChanged:
                         {
                             var modifierFlags = (NSEventModifierMask)Cocoa.SendUint(e, selModifierFlags);
+                            capslockOn = (modifierFlags & NSEventModifierMask.AlphaShiftKeyMask) != 0;
```
Comment (swoolcock) [inline src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs]: If you're going to work with modifier flags, you should put it in `UpdateModifierFlags`.

**Review by swoolcock — CHANGES_REQUESTED** [review]

**Comment by plankp** [comment]
I kept the values altered for `OnKeyPress` because I think that is what the other platform's native windows are doing.

**Comment by swoolcock** [comment]
I misread how `OnKeyPress` works.  You are correct in that it expects the character rather than the key code, as that goes into `OnKeyDown`.  The remaining issue is that while your caps lock check will work, it still does not handle accented key combinations.  Personally I would put that in the same PR, but it's up to you if you want to get the caps lock one in first then make a second one later.

**Comment by plankp** [comment]
I think this should only deal with caps lock keys. (at the moment I really cannot think of how to implement the accented keys...)

**Comment by plankp** [comment]
Turns out key presses that only create a composable accent key will not work. For example layout for "Canadian French - CSA" circumflex marking (mapped to `[` on US keyboard). If you press it, nothing will appear. Don't know how to fix that.

On that note, since option-e on "ABC" layout represents a composable accent key, to keep it consistent, should it also follow the behaviour of `characters` (which results in an empty string)? Or use the current fallback (which cannot handle the circumflex marking example).

It does handle some option keys combinations properly, for example, option-a on "ABC" layout since it produces an actual character instead of a composable accent key.

**Pull request closed without merging**